src/passes/TypeSSA.cpp

```diff
@@ -0,0 +1,257 @@
+    processNews(moduleFinder.news);
+
+    // Modify the ones we found are relevant. We must modify them all at once as
+    // in the isorecursive type system we want to create a single new rec group
+    // for them all (see below).
+    modifyNews();
+  }
+
+  News newsToModify;
+
+  // All the types that are seen in struct.new etc. operations anywhere in the
+  // program. We use this below to check for an error condition regarding rec
+  // groups.
+  std::unordered_set<HeapType> allSeenTypes;
+
+  // As we generate new names, use a consistent index.
+  Index nameCounter = 0;
+
+  void processNews(const News& news) {
+    for (auto* curr : news.structNews) {
+      if (isInteresting(curr)) {
+        newsToModify.structNews.push_back(curr);
+        allSeenTypes.insert(curr->type.getHeapType());
+      }
+    }
+  }
+
+  void modifyNews() {
+    // We collected all the instructions we want to create new types for. Now we
+    // can create those new types, which we do all at once. It is important to
+    // do so in the isorecursive type system as if we create a new singleton rec
+    // group for each then all those will end up identical to each other, so
+    // instead we'll make a single big rec group for them all at once.
+    //
+    // Our goal is to create a completely new/fresh/private type for each
+    // instruction that we found. In the isorecursive type system there isn't an
+    // explicit way to do so, but at least if the new rec group is very large
+    // the risk of collision with another rec group in the program is small.
+    // Note that the risk of collision with things outside of this module is
+    // also a possibility, and so for that reason this pass is likely mostly
+    // useful in the closed-world scenario.
+
+    auto& structNews = newsToModify.structNews;
+    TypeBuilder builder(structNews.size());
+    for (Index i = 0; i < structNews.size(); i++) {
+      auto* curr = structNews[i];
+      auto oldType = curr->type.getHeapType();
+      builder.setHeapType(i, oldType.getStruct());
+      builder.setSubType(i, oldType);
+    }
+    builder.createRecGroup(0, structNews.size());
+    auto result = builder.build();
+    assert(!result.getError());
+    auto newTypes = *result;
+    assert(newTypes.size() == structNews.size());
+
+    // The new types must not overlap with any existing ones. If they do, then
+    // it would be unsafe to apply this optimization (if casts exist to the
+    // existing types, the new types merged with them would now succeed on those
+    // casts). We could try to create a "weird" rec group to avoid this (e.g. we
+    // could make a rec group larger than any existing one, or with an initial
+    // member that is "random"), but hopefully this is rare, so just error for
+    // now.
+    for (auto newType : newTypes) {
+      if (allSeenTypes.count(newType)) {
+        Fatal() << "Rec group collision in TypeSSA! Please file a bug";
+      }
+    }
```

Comment:
Once you switch to using `ModuleUtils::collectHeapTypes`, you will only need to check a single type here rather than checking all of them. Since the types are all in the same rec group, if one of them is collected, they will all be collected.

Also, this seems like it should produce a warning at most and then bail out of the optimization without changing anything rather than producing a fatal error. We could more easily test this case that way, too.

Reply:
Hmm, good point about just checking one. I'll add that.

I do think this should error. A warning + bailout will likely mean this ends up ignored by the user, and we silently miss optimization opportunities. If that's an important case we'd have a slowdown there that is very hard to diagnose if the warning is missed. With an error the user would report a bug and then we'd know we need to prioritize a proper hack for the isorecursive case.

Reply:
Ok, if our response to a bug report about this error would be to implement a workaround, then that sounds good to me.